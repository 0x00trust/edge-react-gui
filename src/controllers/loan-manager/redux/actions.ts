import { add, gt } from 'biggystring'
import { EdgeAccount } from 'edge-core-js'

import { BorrowPlugin } from '../../../plugins/borrow-plugins/types'
import { Dispatch, GetState } from '../../../types/reduxTypes'
import { makeCleanStore } from '../../../util/CleanStore'
import { scheduleActionProgram } from '../../action-queue/redux/actions'
import { ActionProgram } from '../../action-queue/types'
import { borrowPluginMap } from '../borrowPluginConfig'
import { asLoanAccountMapRecord, LOAN_ACCOUNT_MAP, LOAN_MANAGER_STORE_ID, LoanAccountEntry, LoanProgramEdge, LoanProgramType } from '../store'
import { LoanAccount } from '../types'

type SetLoanAccountAction = {
  type: 'LOAN_MANAGER/SET_LOAN_ACCOUNT'
  loanAccount: LoanAccount
}
type DeleteLoanAccountAction = {
  type: 'LOAN_MANAGER/DELETE_LOAN_ACCOUNT'
  id: string
}
type UpdateSyncRatio = {
  type: 'LOAN_MANAGER/SET_SYNC_RATIO'
  syncRatio: number
}

export type LoanManagerActions = SetLoanAccountAction | DeleteLoanAccountAction | UpdateSyncRatio

/**
 * Save a new LoanAccount to disk and sets it in the Redux state.
 * It will throw if specified LoanAccount exists already.
 */
export const createLoanAccount = (loanAccount: LoanAccount) => async (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const account: EdgeAccount = state.core.account
  const store = makeCleanStore(account, LOAN_MANAGER_STORE_ID)
  const loanAccountMapRecord = await store.initRecord(LOAN_ACCOUNT_MAP, asLoanAccountMapRecord)

  if (loanAccountMapRecord.data[loanAccount.id] == null) {
    const { borrowPlugin, borrowEngine, programEdges } = loanAccount
    const loanEntry: LoanAccountEntry = {
      walletId: borrowEngine.currencyWallet.id,
      borrowPluginId: borrowPlugin.borrowInfo.borrowPluginId,
      programEdges
    }

    loanAccountMapRecord.update({ ...loanAccountMapRecord.data, [loanAccount.id]: loanEntry })
  } else {
    throw new Error('Creating duplicate LoanAccount id: ' + loanAccount.id)
  }

  dispatch({
    type: 'LOAN_MANAGER/SET_LOAN_ACCOUNT',
    loanAccount: loanAccount
  })
}

/**
 * Load all LoanAccounts from disk to the redux store and initializes
 * associated BorrowEngines.
 */
export const loadLoanAccounts = (account: EdgeAccount) => async (dispatch: Dispatch, getState: GetState) => {
  const store = makeCleanStore(account, LOAN_MANAGER_STORE_ID)
  const loanAccountMapRecord = await store.initRecord(LOAN_ACCOUNT_MAP, asLoanAccountMapRecord)

  for (const key of Object.keys(loanAccountMapRecord.data)) {
    const record = loanAccountMapRecord.data[key]
    const { walletId, borrowPluginId, programEdges } = record
    const wallet = await account.waitForCurrencyWallet(walletId)
    const borrowPlugin = borrowPluginMap[borrowPluginId]
    const borrowEngine = await borrowPlugin.makeBorrowEngine(wallet)
    const loanAccount: LoanAccount = {
      id: walletId,
      borrowPlugin,
      borrowEngine,
      programEdges
    }
    dispatch({
      type: 'LOAN_MANAGER/SET_LOAN_ACCOUNT',
      loanAccount
    })
  }
}

/**
 * Update an existing LoanAccount.
 * It will throw if specified LoanAccount is not found.
 */
export const updateLoanAccount = (loanAccount: LoanAccount) => async (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const account: EdgeAccount = state.core.account
  const store = makeCleanStore(account, LOAN_MANAGER_STORE_ID)
  const loanAccountMapRecord = await store.initRecord(LOAN_ACCOUNT_MAP, asLoanAccountMapRecord)

  if (loanAccountMapRecord.data[loanAccount.id] == null) {
    throw new Error('Could not find LoanAccount id: ' + loanAccount.id)
  } else {
    const { borrowPlugin, borrowEngine, programEdges } = loanAccount
    const loanEntry: LoanAccountEntry = {
      walletId: borrowEngine.currencyWallet.id,
      borrowPluginId: borrowPlugin.borrowInfo.borrowPluginId,
      programEdges
    }

    loanAccountMapRecord.update({ ...loanAccountMapRecord.data, [loanAccount.id]: loanEntry })
  }

  dispatch({
    type: 'LOAN_MANAGER/SET_LOAN_ACCOUNT',
    loanAccount: loanAccount
  })
}

/**
 * Delete an existing LoanAccount.
 * It will throw if specified LoanAccount is not found.
 */
export const deleteLoanAccount = (loanAccountOrId: LoanAccount | string) => async (dispatch: Dispatch, getState: GetState) => {
  const state = getState()
  const account: EdgeAccount = state.core.account
  const store = makeCleanStore(account, LOAN_MANAGER_STORE_ID)
  const loanAccountMapRecord = await store.initRecord(LOAN_ACCOUNT_MAP, asLoanAccountMapRecord)

  const loanAccountId = typeof loanAccountOrId === 'string' ? loanAccountOrId : loanAccountOrId.id
  if (loanAccountMapRecord.data[loanAccountId] != null) {
    const { [loanAccountId]: _, ...loanAccountMapRecordUpdated } = loanAccountMapRecord.data
    loanAccountMapRecord.update(loanAccountMapRecordUpdated)
  } else {
    throw new Error('Could not find LoanAccount id: ' + loanAccountId)
  }

  dispatch({
    type: 'LOAN_MANAGER/DELETE_LOAN_ACCOUNT',
    id: loanAccountId
  })
}

/**
 * Execute action program, and update associated LoanAccount.
 */
export const runLoanActionProgram =
  (loanAccount: LoanAccount, actionProgram: ActionProgram, programType: LoanProgramType) =>
  async (dispatch: Dispatch, getState: GetState): Promise<LoanAccount> => {
    await dispatch(scheduleActionProgram(actionProgram))
    const programEdge: LoanProgramEdge = {
      programId: actionProgram.programId,
      programType
    }
    await dispatch(updateLoanAccount({ ...loanAccount, programEdges: [...loanAccount.programEdges, programEdge] }))

    return loanAccount
  }

export const resyncLoanAccounts = (account: EdgeAccount) => async (dispatch: Dispatch, getState: GetState) => {
  const store = makeCleanStore(account, LOAN_MANAGER_STORE_ID)
  const loanAccountMapRecord = await store.initRecord(LOAN_ACCOUNT_MAP, asLoanAccountMapRecord)

  const typeHack: any = Object.values(borrowPluginMap)
  const borrowPlugins: BorrowPlugin[] = typeHack

  // `loanAccountIds` is synonymous to `walletIds`
  const ids = Object.keys(account.currencyWallets)

  dispatch({
    type: 'LOAN_MANAGER/SET_SYNC_RATIO',
    syncRatio: 0
  })

  let progress = 0
  const promises = ids
    .map(async (id, i) => {
      const loanAccountId = id
      try {
        const wallet = await account.waitForCurrencyWallet(loanAccountId)

        // Find plugin which support currency wallet
        const currencyPluginId = wallet.currencyConfig.currencyInfo.pluginId
        const borrowPlugin = borrowPlugins.find(borrowPlugin => borrowPlugin.borrowInfo.currencyPluginId === currencyPluginId)

        // Skip wallets that don't have a supported borrow plugin
        if (borrowPlugin == null) return

        // Make borrow engine
        const borrowEngine = await borrowPlugin.makeBorrowEngine(wallet)

        // Wait for borrow engine to be fully synced
        await new Promise(resolve => {
          borrowEngine.watch('syncRatio', syncRatio => {
            // @ts-expect-error
            if (syncRatio >= 1) resolve()
          })
        })

        const hasDebt = gt(
          borrowEngine.debts.reduce((sum, debt) => add(sum, debt.nativeAmount), '0'),
          '0'
        )
        const hasCollateral = gt(
          borrowEngine.collaterals.reduce((sum, collateral) => add(sum, collateral.nativeAmount), '0'),
          '0'
        )

        if (hasDebt && hasCollateral) {
          // Ignore creating account if already created
          if (loanAccountMapRecord.data[loanAccountId]) return
          // Create loan account if not already created
          const loanAccount: LoanAccount = {
            id: loanAccountId,
            borrowPlugin,
            borrowEngine,
            programEdges: []
          }
          await dispatch(createLoanAccount(loanAccount))
        } else {
          // Remove loans "emptied out" loans
          const existingLoanAccount = loanAccountMapRecord.data[loanAccountId]
          if (existingLoanAccount != null) {
            await dispatch(deleteLoanAccount(loanAccountId))
          }
        }
      } catch (error: any) {
        console.error(error)
      }
    })
    .map(async promise =>
      promise.then(() => {
        const syncRatio = ++progress / ids.length
        // console.log('\n###\n', { total: ids.length, progress, syncRatio }, '\n###\n')
        dispatch({
          type: 'LOAN_MANAGER/SET_SYNC_RATIO',
          syncRatio
        })
      })
    )

  await Promise.all(promises)

  dispatch({
    type: 'LOAN_MANAGER/SET_SYNC_RATIO',
    syncRatio: 1
  })
}
