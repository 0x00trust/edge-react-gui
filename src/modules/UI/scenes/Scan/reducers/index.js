// @flow

import { addressModalVisible } from './addressModalVisible.js'
import { parsedUri } from './parsedUri.js'
import { scanEnabled } from './scanEnabled.js'
import { torchEnabled } from './torchEnabled.js'

export { addressModalVisible, scanEnabled, torchEnabled, parsedUri }
