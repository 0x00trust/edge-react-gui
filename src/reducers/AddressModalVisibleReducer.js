// @flow

import type { Action } from '../modules/ReduxTypes.js'

export const initialState = false
export type State = boolean
export const addressModalVisibleReducer = (state: State = initialState, action: Action) => {
  switch (action.type) {
    case 'SAVE_EDGE_LOBBY':
    case 'SET_LOBBY_ERROR': {
      return false
    }

    case 'TOGGLE_ADDRESS_MODAL_VISIBILITY': {
      return !state
    }

    default:
      return state
  }
}
