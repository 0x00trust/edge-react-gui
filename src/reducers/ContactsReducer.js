// @flow

import type { Action } from '../types/reduxTypes.js'
import type { GuiContact } from '../types/types.js'

export type ContactsState = Array<GuiContact>

export const initialState = []

export const contacts = (state: ContactsState = initialState, action: Action): ContactsState => {
  switch (action.type) {
    case 'CONTACTS/LOAD_CONTACTS_SUCCESS': {
      if (!action.data) throw new Error('Invalid action')
      return action.data.contacts
    }

    default:
      return state
  }
}
