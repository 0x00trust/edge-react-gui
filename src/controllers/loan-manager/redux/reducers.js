// @flow

import { type Reducer, combineReducers } from 'redux'

import { type Action } from '../../../types/reduxTypes'
import { type LoanAccountMap } from '../types'

export type LoanManagerState = {
  +loanAccounts: LoanAccountMap
}

export const loanManager: Reducer<LoanManagerState, Action> = combineReducers({
  loanAccounts(state: LoanAccountMap = {}, action: Action): LoanAccountMap {
    switch (action.type) {
      case 'LOAN_MANAGER/SET_LOAN_ACCOUNT': {
        const { loanAccount } = action
        return {
          ...state,
          [loanAccount.id]: loanAccount
        }
      }
      case 'LOAN_MANAGER/DELETE_LOAN_ACCOUNT': {
        const { id } = action
        const { [id]: _, ...remainingState } = state
        return remainingState
      }
      case 'LOGOUT': {
        return {}
      }
      default:
        return state
    }
  }
})
