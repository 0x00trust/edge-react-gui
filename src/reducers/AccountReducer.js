// @flow

import { type Reducer, combineReducers } from 'redux'

import { type CreationReason } from '../types/CreationReason.js'
import { type Action } from '../types/reduxTypes.js'

/**
 * App state tied to the core account object.
 */
export type AccountState = {
  +creationReason: CreationReason | null
  // TODO: Move account settings in here
}

const accountInner: Reducer<AccountState, Action> = combineReducers({
  creationReason (state: CreationReason | null = null, action: Action): CreationReason | null {
    return action.type === 'CREATION_REASON_LOADED' ? action.data : state
  }
})

// Shared logout logic:
export const account: Reducer<AccountState, Action> = (state: AccountState | void, action: Action) => {
  if (action.type === 'LOGOUT' || action.type === 'DEEP_LINK_RECEIVED') {
    return accountInner(undefined, { type: 'DUMMY_ACTION_PLEASE_IGNORE' })
  }
  return accountInner(state, action)
}
