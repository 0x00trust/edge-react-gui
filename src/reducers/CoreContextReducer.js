// @flow

import { type DiskletFolder } from 'disklet'
import type { EdgeContext } from 'edge-core-js'
import { type Reducer } from 'redux'

import { type Action } from '../types/reduxTypes.js'

export type ContextState = {
  context: EdgeContext | Object,
  folder: DiskletFolder | Object
}

const initialState = {
  context: {},
  folder: {}
}

export const context: Reducer<ContextState, Action> = (state = initialState, action: Action) => {
  switch (action.type) {
    case 'CORE/CONTEXT/ADD_CONTEXT': {
      if (!action.data) throw new Error('Invalid action')
      const context: EdgeContext = action.data.context
      const folder: DiskletFolder = action.data.folder
      return {
        ...state,
        context,
        folder
      }
    }

    default:
      return state
  }
}
