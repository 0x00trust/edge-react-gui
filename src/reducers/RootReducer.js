// @flow

import { type Reducer, combineReducers } from 'redux'

import { type ExchangeRatesState, exchangeRates } from '../modules/ExchangeRates/reducer.js'
import { type UiState, ui } from '../modules/UI/reducer.js'
import { type Action } from '../types/reduxTypes.js'
import { type ContactsState, contacts } from './ContactsReducer.js'
import { type CoreState, core } from './CoreReducer.js'
import { type CryptoExchangeState, cryptoExchange } from './CryptoExchangeReducer.js'
import { type PermissionsState, permissions } from './PermissionsReducer.js'

export type RootState = {
  // Next username to auto-fill at the login screen, or blank if none:
  +nextUsername: string | null,

  // Nested reducers:
  +contacts: ContactsState,
  +core: CoreState,
  +cryptoExchange: CryptoExchangeState,
  +exchangeRates: ExchangeRatesState,
  +permissions: PermissionsState,
  +ui: UiState
}

export const rootReducer: Reducer<RootState, Action> = combineReducers({
  nextUsername (state: string | null = null, action: Action): string | null {
    switch (action.type) {
      case 'DEEP_LINK_RECEIVED':
        return null
      case 'LOGOUT': {
        const { username = null } = action.data
        return username
      }
    }
    return state
  },

  // Nested reducers:
  contacts,
  core,
  cryptoExchange,
  exchangeRates,
  permissions,
  ui
})
