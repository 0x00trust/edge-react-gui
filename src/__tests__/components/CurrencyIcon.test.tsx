import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import { createStore } from 'redux'

import { CryptoIcon } from '../../components/icons/CryptoIcon'
import { rootReducer } from '../../reducers/RootReducer'

describe('CryptoIcon', () => {
  const mockState: any = {
    core: {
      account: {
        currencyWallets: {
          '332s0ds39f': {
            pluginId: 'bitcoin',
            watch: () => {}
          }
        }
      }
    }
  }
  const store = createStore(rootReducer, mockState)

  it('should render with loading props', () => {
    const actual = renderer.create(
      <Provider store={store}>
        <CryptoIcon pluginId="bitcoin" tokenId="bitcoin" walletId="332s0ds39f" marginRem={1} />
      </Provider>
    )

    expect(actual.toJSON()).toMatchSnapshot()
  })
})
