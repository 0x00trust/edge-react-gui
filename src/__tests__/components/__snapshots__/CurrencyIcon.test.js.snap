// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyIcon should render with loading props 1`] = `
<View
  style={
    Object {
      "height": 45,
      "marginBottom": 22,
      "marginLeft": 22,
      "marginRight": 22,
      "marginTop": 22,
      "width": 45,
    }
  }
>
  <WalletSyncCircle
    edgeWallet={
      Object {
        "pluginId": "bitcoin",
        "watch": [Function],
      }
    }
    size={45}
  />
  <FastImage
    source={
      Object {
        "uri": "https://content.edge.app/currencyIcons/bitcoin/bitcoin.png",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;
